presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregationIntoTableScan.java

```diff
@@ -218,4 +232,82 @@ private AggregateFunction toAggregateFunction(Context context, AggregationNode.A
                 aggregation.isDistinct(),
                 filter);
     }
+
+    public static class PushAggregationIntoTableScanWithoutGroupingSets
+            extends PushAggregationIntoTableScan
+    {
+        private static final Pattern<AggregationNode> AGGREGATION_WITHOUT_GROUPING_SETS =
+                aggregation()
+                        .matching(PushAggregationIntoTableScan::allArgumentsAreSimpleReferences)
+                        .matching(PushAggregationIntoTableScan::hasNoMasks)
+                        .with(source().matching(tableScan().capturedAs(TABLE_SCAN)));
+
+        public PushAggregationIntoTableScanWithoutGroupingSets(Metadata metadata)
+        {
+            super(metadata);
+        }
+
+        @Override
+        public Pattern<AggregationNode> getPattern()
+        {
+            return AGGREGATION_WITHOUT_GROUPING_SETS;
+        }
+
+        @Override
+        public List<List<Symbol>> groupBySymbols(AggregationNode node, Captures captures)
+        {
+            return List.of(node.getGroupingSets().getGroupingKeys());
+        }
+
+        @Override
+        protected Map<Symbol, Symbol> groupingSetSymbolMapping(AggregationNode node, Captures captures)
+        {
+            return node.getGroupingSets().getGroupingKeys().stream()
+                    .collect(Collectors.toMap(identity(), identity()));
+        }
+    }
+
+    public static class PushAggregationIntoTableScanWithGroupingSets
+            extends PushAggregationIntoTableScan
+    {
+        private static final Capture<GroupIdNode> GROUP_ID = newCapture();
+        private static final Pattern<AggregationNode> AGGREGATION_WITH_GROUPING_SETS =
+                aggregation()
+                        .matching(PushAggregationIntoTableScan::allArgumentsAreSimpleReferences)
+                        .matching(PushAggregationIntoTableScan::hasNoMasks)
+                        .with(source().matching(typeOf(GroupIdNode.class).capturedAs(GROUP_ID)
+                                .with(source().matching(tableScan().capturedAs(TABLE_SCAN)))));
+
+        public PushAggregationIntoTableScanWithGroupingSets(Metadata metadata)
+        {
+            super(metadata);
+        }
+
+        @Override
+        public Pattern<AggregationNode> getPattern()
+        {
+            return AGGREGATION_WITH_GROUPING_SETS;
+        }
+
+        @Override
+        public List<List<Symbol>> groupBySymbols(AggregationNode node, Captures captures)
+        {
+            GroupIdNode groupIdNode = captures.get(GROUP_ID);
+
+            return groupIdNode.getGroupingSets().stream()
+                    .map(groupIdNodeSymbols -> groupIdNodeSymbols.stream()
+                            .map(groupIdNode.getGroupingColumns()::get)
+                            .collect(toList())
+                    ).collect(toList());
```

Comment:
put `)` on prev line and use `toImmutableList`

Reply:
Done.